### PR TITLE
Struct migrations from canary checker

### PR DESCRIPTION
**Pull request by adityathebe**

### CODE REVIEW

File: models/checks.go [hunk 1]
```diff
@@ -0,0 +1,849 @@
...
+import (
+	"encoding/json"
+	"fmt"
+	"strings"
+
+	"github.com/flanksource/kommons"
+	v1 "k8s.io/api/core/v1"
+)
+
+type ICheck interface {
```
Comment (adityathebe) [inline models/checks.go:12]: @yashmehrotra had to rename this interface from `Check`. It was in `canary-checker/pkg/api/external`

File: models/components.go [hunk 2]
```diff
@@ -0,0 +1,539 @@
...
+	CreatedAt        time.Time  `gorm:"column:created_at" json:"created_at,omitempty"`
+	UpdatedAt        time.Time  `gorm:"column:updated_at" json:"updated_at,omitempty"`
+	DeletedAt        *time.Time `gorm:"column:deleted_at" json:"deleted_at,omitempty"`
+}
+
+func (cr ComponentRelationship) TableName() string {
+	return "component_relationships"
+}
+
+type ConfigComponentRelationship struct {
```
Comment (adityathebe) [inline models/components.go:528]: This came from `canary-checker/pkg/db`. Added the `gorm` tags & the last 3 fields.

**Comment by adityathebe** [comment]
Most of what's in here is just copy-paste. For the few things where I had to intervene, I've left some comments.

File: models/canary.go [hunk 3]
```diff
@@ -0,0 +1,274 @@
...
+// 		Source:    canary.Annotations["source"],
+// 		Schedule:  canary.Spec.GetSchedule(),
+// 		Checks:    types.JSONStringMap(checks),
+// 	}, nil
+// }
+
+type ResultMode string
+
+// CanarySpec defines the desired state of Canary
+type CanarySpec struct {
```
Comment (yashmehrotra) [inline models/canary.go:81]: This can be removed

Comment (yashmehrotra) [reply models/canary.go:81]: Everything below this can be

File: models/checks.go [hunk 4]
```diff
@@ -0,0 +1,849 @@
+package models
```
Comment (yashmehrotra) [inline models/checks.go:1]: I think this entire file can be removed

File: models/components.go [hunk 5]
```diff
@@ -0,0 +1,539 @@
...
+	Configs          Configs                  `json:"configs,omitempty" gorm:"type:configs"`
+	CostPerMinute    float64                  `json:"cost_per_minute,omitempty" gorm:"column:cost_per_minute"`
+	CostTotal1d      float64                  `json:"cost_total_1d,omitempty" gorm:"column:cost_total_1d"`
+	CostTotal30d     float64                  `json:"cost_total_30d,omitempty" gorm:"column:cost_total_30d"`
+	CostTotal7d      float64                  `json:"cost_total_7d,omitempty" gorm:"column:cost_total_7d"`
+	CreatedAt        time.Time                `json:"created_at,omitempty" time_format:"postgres_timestamp"`
+	DeletedAt        *time.Time               `json:"deleted_at,omitempty" time_format:"postgres_timestamp" swaggerignore:"true"`
+	ExternalId       string                   `json:"external_id,omitempty"` //nolint
+	Icon             string                   `json:"icon,omitempty"`
+	ID               uuid.UUID                `json:"id,omitempty" gorm:"default:generate_ulid()"` //nolint
```
Comment (yashmehrotra) [inline models/components.go:36]: Keep ID as the first, looks better

File: models/components.go (hunk 2 above)
Comment (yashmehrotra) [reply models/components.go:528]: Gorm tags might not be required since the tags we added are the default anyway

File: models/db_types.go [hunk 6]
```diff
@@ -0,0 +1,11 @@
+package models
```
Comment (yashmehrotra) [inline models/db_types.go:1]: Move this to `types/types.go`

File: models/varsource.go [hunk 7]
```diff
@@ -0,0 +1,164 @@
...
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	runtime "k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/apimachinery/pkg/util/validation"
+	"k8s.io/client-go/kubernetes"
+)
+
+// VarSource represents a source for a value
+type VarSource struct {
```
Comment (yashmehrotra) [inline models/varsource.go:18]: This is a k8s entity, might not be required

**Review by yashmehrotra — CHANGES_REQUESTED** [review]
Might have to delete more things from here

File: models/components.go [hunk 8]
```diff
@@ -0,0 +1,539 @@
...
+	"gorm.io/gorm"
+	"gorm.io/gorm/clause"
+	"gorm.io/gorm/schema"
+)
+
+type Components []*Component
+
+type Component struct {
+	Checks           Checks                   `json:"checks,omitempty" gorm:"-"`
+	ComponentChecks  ComponentChecks          `json:"-" gorm:"column:component_checks" swaggerignore:"true"`
```
Comment (yashmehrotra) [inline models/components.go:24]: @moshloop Do we have a need for this field ?

File: models/config.go [hunk 9]
```diff
@@ -1,16 +1,79 @@
...
 
 	"github.com/flanksource/duty/types"
+	"github.com/google/uuid"
 	"github.com/lib/pq"
 	"gorm.io/gorm"
+	"gorm.io/gorm/clause"
+	"gorm.io/gorm/schema"
 )
 
+type Config struct {
```
Comment (yashmehrotra) [inline models/config.go:20]: Use ConfigItem instead

File: models/components.go (hunk 5 above)
Comment (adityathebe) [reply models/components.go:36]: okay. I generally sort fields in a relatively large struct so it's easier to scan

File: models/config.go (hunk 9 above)
Comment (adityathebe) [reply models/config.go:20]: @yashmehrotra Does this mean `models.Config` can be deleted?

File: models/checks.go (hunk 4 above)
Comment (adityathebe) [reply models/checks.go]: It's removed

File: models/config.go (hunk 9 above)
Comment (yashmehrotra) [reply models/config.go:20]: Yeah, they refer to the same thing

File: types/types.go [hunk 10]
```diff
@@ -17,6 +17,17 @@ import (
...
+	SQLServerType = "sqlserver"
+	PostgresType  = "postgres"
+	SqliteType    = "sqlite"
+	MysqlType     = "mysql"
+	Text          = "TEXT"
+	JSONType      = "JSON"
+	JSONBType     = "JSONB"
+	NVarcharType  = "NVARCHAR(MAX)"
+)
+
```
Comment (yashmehrotra) [inline types/types.go:30]: 👍🏽

File: models/components.go [hunk 11]
```diff
@@ -0,0 +1,482 @@
...
+	Unit string `json:"unit,omitempty"`
+	Max  *int64 `json:"max,omitempty"`
+	Min  int64  `json:"min,omitempty"`
+
+	Status         string `json:"status,omitempty"`
+	LastTransition string `json:"lastTransition,omitempty"`
+	Links          []Link `json:"links,omitempty"`
+}
+
+func (p Property) GetValue() interface{} {
```
Comment (yashmehrotra) [inline models/components.go:180]: prefer `any` instead of `interface{}`

File: go.mod [hunk 12]
```diff
@@ -4,8 +4,8 @@ go 1.19
 
 require (
 	ariga.io/atlas v0.9.0
+	github.com/fergusstrange/embedded-postgres v1.19.0
```
Comment (yashmehrotra) [inline go.mod:7]: which package is requiring this ?

File: models/components.go [hunk 13]
```diff
@@ -0,0 +1,482 @@
...
+	}
+	return ""
+}
+
+func (p Properties) GormValue(ctx context.Context, db *gorm.DB) clause.Expr {
+	data, _ := json.Marshal(p)
+	return gorm.Expr("?", data)
+}
+
+type Check struct {
```
Comment (yashmehrotra) [inline models/components.go:330]: can move this to models/checks.go

File: go.mod (hunk 12 above)
Comment (adityathebe) [reply go.mod:7]: `schema_test.go`

File: models/config.go (hunk 9 above)
Comment (adityathebe) [reply models/config.go]: okay, deleted it.

**Review by yashmehrotra — APPROVED** [review]

**Pull request merged**